src/lib/comms/sol-network-impl-unix-generic.c

```diff
@@ -0,0 +1,140 @@
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <limits.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netdb.h>
+#include <unistd.h>
+
+#define SOL_LOG_DOMAIN &_log_domain
+#include "sol-log-internal.h"
+#include "sol-macros.h"
+#include "sol-mainloop.h"
+#include "sol-network.h"
+#include "sol-network-util.h"
+#include "sol-util-internal.h"
+#include "sol-util-file.h"
+
+SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "network");
+
+struct sol_network {
+    unsigned int count;
+    int seq;
+
+    struct sol_timeout *hostname_worker;
+    struct sol_ptr_vector hostname_handles;
+};
+
+static struct sol_network *network = NULL;
+
+#include "sol-network-impl-unix-common.c"
+
+SOL_API int
+sol_network_init(void)
+{
+    SOL_LOG_INTERNAL_INIT_ONCE;
+    return 0;
+
+    if (network != NULL) {
+        network->count++;
+        return 0;
+    }
+
+    network = calloc(1, sizeof(*network));
+    SOL_NULL_CHECK(network, -ENOMEM);
+
+    sol_ptr_vector_init(&network->hostname_handles);
+
+    network->seq = 0;
+    network->count = 1;
+    return 0;
+}
+
+SOL_API void
+sol_network_shutdown(void)
+{
+    uint16_t idx;
+    struct sol_network_hostname_handle *ctx;
+
+    if (!network)
+        return;
+
+    if (network->hostname_worker)
+        sol_timeout_del(network->hostname_worker);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&network->hostname_handles, ctx, idx)
+        hostname_handle_free(ctx);
+    sol_ptr_vector_clear(&network->hostname_handles);
+
+    free(network);
+    network = NULL;
+}
+
+SOL_API bool
+sol_network_subscribe_events(void (*cb)(void *data, const struct sol_network_link *link,
+    enum sol_network_event event),
+    const void *data)
+{
```

Comment:
can we find a way to do this in other platforms? Otherwise if network links are enabled/disabled the servers such as CoAP will not work properly, like joining multicast groups and so on.